rust/eos-rs/src/api/overlay.rs

```diff
@@ -0,0 +1,112 @@
+//! Module for dealing with the load-state of Nintendo DS overlays.
+
+use crate::ffi;
+
+/// This represents a promise to the compiler that while a reference
+/// to this struct exists, the overlay N is loaded.
+///
+/// This is used as a parameter in functions that need a given overlay to be loaded.
+///
+/// Dropping the lease does *not* unload an overlay, but you can use
+/// the [`Self::unload`] method on a lease to do so.
+///
+/// You can clone a lease, but you will have the responsibility to make
+/// sure the overlay is not unloaded while clone lease is in use too.
+#[derive(Clone)]
+pub struct OverlayLoadLease<const N: u32>(());
+
+impl<const N: u32> OverlayLoadLease<N> {
+    /// Gets a lease on a loaded overlay.
+    ///
+    /// This is a runtime checked version of `acquire_unchecked`.
+    /// It will panic if the overlay is not loaded.
+    ///
+    /// # Important
+    /// Even though this function is marked as safe, since it
+    /// checks if the overlay is loaded, you still need to make
+    /// sure the overlay is not unloaded during the lifetime
+    /// of the returned object.
+    pub fn acquire() -> Self {
+        if !Self::is_loaded() {
+            panic!("Overlay {} is not loaded.", N);
+        }
+        Self(())
+    }
+    /// Gets a lease on a loaded overlay.
+    ///
+    /// # Safety
+    /// You need to make sure the overlay is actually loaded and won't
+    /// get unloaded during the lifetime of the returned object.
+    pub unsafe fn acquire_unchecked() -> Self {
+        Self(())
+    }
+
+    /// Checks if the overlay is loaded.
+    pub fn is_loaded() -> bool {
+        unsafe { ffi::OverlayIsLoaded(N) > 0 }
```

Comment:
This isn't right. `OverlayIsLoaded`, `LoadOverlay`, and `UnloadOverlay` all use `enum overlay_group_id`, not the actual overlay ID. So with this implementation, `OverlayLoadLead<29>` will correspond to overlay 25 (`OGROUP_OVERLAY_25 = 29`).

Reply:
Yep, just noticed this today too while testing. Will rewrite it to match against the enum.